include/seqan3/alignment/pairwise/policy/alignment_matrix_policy.hpp

```diff
@@ -0,0 +1,146 @@
+     * \param[in] snd_range The second range (corresponds to vertical dimension).
+     * \param[in] band The band used to initialise the matrices.
+     * \param[in] cache The current cache used by the alignment algorithm.
+     *
+     * \details
+     *
+     * Initialises the underlying banded score and trace matrices and sets the respective matrix iterators to the begin
+     * of the corresponding matrix. Using the additional band parameter the actual dimensions are reduced according
+     * to the matrix implementation. For the banded case one additional cell per column is stored such that we can read
+     * from it without introducing a case distinction inside of the algorithm implementation. However, this cell needs
+     * to be properly initialised with an infinity value. To emulate the infinity for integral values we use the
+     * smallest representable value and subtract the gap extension score (assumed to be always negative) from it.
+     * In the algorithm we never write to this cell and only add the extension costs to the read value. This way we
+     * can get the smallest possible value as an infinity.
+     */
+    template <std::ranges::forward_range fst_range_t, std::ranges::forward_range snd_range_t, typename score_t>
+    constexpr void allocate_matrix(fst_range_t && fst_range,
+                                   snd_range_t && snd_range,
+                                   static_band const & band,
+                                   alignment_algorithm_cache<score_t> const & cache)
+    {
+        assert(cache.gap_extension_score <= 0); // We expect it to be always non-positive.
+
+        score_t inf = std::numeric_limits<score_t>::min() - cache.gap_extension_score;
+        score_matrix = score_matrix_type{fst_range, snd_range, band, inf};
+        trace_matrix = trace_matrix_type{fst_range, snd_range, band};
+
+        initialise_matrix_iterator();
+    }
+
+    //!\brief Initialises the score and trace matrix iterator after allocating the matrices.
+    constexpr void initialise_matrix_iterator() noexcept
+    {
+        score_matrix_iter = score_matrix.begin();
+        trace_matrix_iter = trace_matrix.begin();
+    }
+
+    /*!\brief Trims the sequences to the band parameters.
+    * \tparam fst_range_t The type of the first sequence (or packed sequences).
+    * \tparam snd_range_t The type of the second sequence (or packed sequences).
+    * \tparam band_t The type of the band.
+    * \param[in] fst_range The first sequence (or packed sequences).
+    * \param[in] snd_range The first sequence (or packed sequences).
+    * \param[in] band The band.
+    *
+    * \details
+    *
+    * If the band does not intersect with the origin or the sink of the matrix the
+    * sequences are trimmed, such that the band starts in the origin and ends in the sink.
+    */
+    template <std::ranges::forward_range fst_range_t, std::ranges::forward_range snd_range_t, typename band_t>
```

Comment:
as above, internally I would presumably not use any concepts in the future. These things are anyway strongly coupled with the algorithm and should be handled before already.

Reply:
Hm... I see that we need to reduce compile time but restricting a type with a concept also helps the reader of the code to better understand what's happening. I first thought here that `band` just is a size and then it suddenly had lower and upper bounds. It is also hard to know if something is checked upstream if you have not developed the whole alignment design :( so should we really sacrifice expressiveness for compile times?

Reply:
If a user needs to wait 5 or minutes for the program to compile then this becomes a bottleneck. Since there is only one band object at the moment I changed it to the explicit class

Reply:
But then we are back to the seqan2 style where everything is just `typename` in every detail function and no one, except the designer of the module, knows which functions get which type of values :(

And I would strongly hope that the compilers will at some point cach concept checks so additional checks do not add overhead.